k8s/dialer.go

```diff
@@ -0,0 +1,93 @@
+package k8s
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"strings"
+
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+// ForwardedPort holds the remote and local mapped port.
+type ForwardedPort struct {
+	Local  uint16
+	Remote uint16
+}
+
+// PortForwardParameters are the needed parameters to call PortForward.
+// Ports value follow the kubectl syntax: <local-port>:<remote-port>
+// 5000 means 5000:5000 listening on 5000 port locally, forwarding to 5000 in the pod
+// 8888:5000 means listening on 8888 port locally, forwarding to 5000 in the pod
+// 0:5000 means listening on a random port locally, forwarding to 5000 in the pod
+// :5000 means listening on a random port locally, forwarding to 5000 in the pod
+type PortForwardParameters struct {
+	Namespace string
+	Pod       string
+	Ports     []string
+	Addresses []string
+}
+
+// PortForwardResult are the ports that have been forwarded.
+type PortForwardResult struct {
+	ForwardedPorts []ForwardedPort
+}
+
+// PortForward executes in a goroutine a port forward command.
+// To kill the thread, you should use the context by cancelling it.
+func (c *Client) PortForward(ctx context.Context, p PortForwardParameters) (*PortForwardResult, error) {
+	req := c.Clientset.CoreV1().RESTClient().Post().Namespace(p.Namespace).
+		Resource("pods").Name(p.Pod).SubResource(strings.ToLower("PortForward"))
+
+	roundTripper, upgrader, err := spdy.RoundTripperFor(c.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, req.URL())
+	stopChan, readyChan := make(chan struct{}, 1), make(chan struct{}, 1)
+	if len(p.Addresses) == 0 {
+		p.Addresses = []string{"localhost"}
+	}
+
+	pw, err := portforward.NewOnAddresses(dialer, p.Addresses, p.Ports, stopChan, readyChan, os.Stdout, os.Stderr)
```

Comment:
I think the output of Stdout and Stderr are only useful when debugging. To avoid this pouting the output, we should make those two `io.Writers` a parameter in `PortForwardParameters`.

And then in `collectPrometheusMetrics`, implement the `io.Writer` interface on top of the logging infrastructure in `connectivity/check/logging.go`. For example:

```
type debugWriter struct {
   ct *ConnectivityTest
}

func (d *debugWriter) Write(b []byte) (int, error) {
   d.ct.Debug(string(b))
   return len(b), nil
}

type warnWriter struct {
   ct *ConnectivityTest
}

func (w *warnWriter) Write(b []byte) (int, error) {
   w.ct.Warn(string(b))
   return len(b), nil
}
```

And then pass in a debugWriter instead of stdout, and a warnWriter instead of stderr.

Reply:
Fixed but I still have a log from the k8s go client port forward library:
https://github.com/kubernetes/client-go/blob/v0.27.1/tools/portforward/portforward.go#L312

The line of code:
```
runtime.HandleError(fmt.Errorf("error accepting connection on port %d: %v", port.Local, err))
```

The log (and there some lines...):
```
E0508 15:06:18.542005   72309 portforward.go:417] error closing listener: close tcp4 127.0.0.1:64508: use of closed network connection
```

I would be happy to discuss how to handle it.

Reply:
Fixed.